src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -85,101 +92,92 @@ trait Action extends DebugEnhancedLogging {
               footer = "Due to these errors in the preconditions, nothing was done."),
             cause = e))
       }
-      _ <- innerExecute().recoverWith {
-        case NonFatal(e) =>
-          List(Failure(e), innerRollback())
-            .collectResults
-            .recoverWith {
-              case e: CompositeException => Failure(ActionRunFailedException(
-                report = generateReport(
-                  header = "Failures during the execute phase:",
-                  throwable = e,
-                  footer = "The actions that were already performed, were rolled back."),
-                cause = e))
-            }
+      t <- innerExecute(a).recoverWith {
+        case e1@CompositeException(es) =>
+          rollback() match {
+            case Success(_) => reportFailure(e1)
+            case Failure(CompositeException(es2)) => reportFailure(CompositeException(es ++ es2))
+            case Failure(e2) => reportFailure(CompositeException(es ++ List(e2)))
+          }
+        case NonFatal(e1) =>
+          rollback() match {
+            case Success(_) => reportFailure(e1)
+            case Failure(CompositeException(es2)) => reportFailure(CompositeException(List(e1) ++ es2))
+            case Failure(e2) => reportFailure(CompositeException(List(e1, e2)))
+          }
       }
-    } yield ()
+    } yield t
```

Comment:
This one is very useful!!! I need to learn to do this more often! Keep reminding me 😃